ext/mbstring/libmbfl/filters/mbfilter_utf16.c

```diff
@@ -1043,3 +1046,89 @@ static void mb_wchar_to_utf16le_avx2(uint32_t *in, size_t len, mb_convert_buf *b
 }
 
 #endif /* defined(ZEND_INTRIN_AVX2_NATIVE) || defined(ZEND_INTRIN_AVX2_RESOLVER) */
+
+static zend_string* mb_cut_utf16be(unsigned char *str, size_t from, size_t len, unsigned char *end)
+{
+	if (len > end - (str + from)) {
+		len = end - (str + from);
+	}
+	from &= ~1;
+	len &= ~1;
+	unsigned char *start = str + from;
+	if (len < 2 || (end - start) < 2) {
+		return zend_empty_string;
+	}
+	/* Check if 1st codepoint is 2nd part of surrogate pair */
+	if (from > 0) {
+		uint32_t start_cp = (*start << 8) + *(start + 1);
+		if (start_cp >= 0xDC00 && start_cp <= 0xDFFF) {
+			uint32_t preceding_cp = (*(start - 2) << 8) + *(start - 1);
+			if (preceding_cp >= 0xD800 && preceding_cp <= 0xDBFF) {
+				from -= 2;
+			}
+		}
+	}
+	/* Same for ending cut point */
+	unsigned char *_end = start + len;
+	if (_end > end) {
+		_end = end;
+	}
+	uint32_t ending_cp = (*(_end - 2) << 8) + *(_end - 1);
+	if (ending_cp >= 0xD800 && ending_cp <= 0xDBFF) {
+		_end -= 2;
+	}
```

Comment:
Isn't the issue for the end if the last codepoint is the 1st part of surrogate pair? In which case shouldn't the end include the second pair too? Or am I not understanding the behaviour of mb_strcut() ?

Reply:
Indeed, I found different behavior
```
var_dump(bin2hex(mb_strcut("\xde\x30\xd8\x44", 0, 4, "UTF-16BE")));
string(4) "de30"
```

3v4l: https://3v4l.org/DoZh7
> string(4) "003f"

Reply:
@Girgias, again, `mb_strcut` is supposed to adjust the cut points (by moving them leftwards) so they fall on character boundaries.

For UTF-16, this can *usually* be accomplished by just rounding odd offsets down to an even offset. The *only* exception is when the resulting cut point falls in the middle of a surrogate pair. These are the only cases where a UTF-16 character takes 4 bytes instead of 2.

A surrogate pair consists of one 16-bit value from 0xD800-DBFF, followed by another 16-bit value from 0xDC00-DFFF.

So to check if our cut point is falling the middle of a surrogate pair, we can check the 16-bit value just before the cut point and see if it is 0xD800-DBFF. (If we wanted to be more thorough, we could also check if the 16-bit value *after* the cut point was 0xDC00-DFFF.) In that case, we need to move the cut point 2 places to the left for it to fall on a "character boundary".

Reply:
@youkidearitai, yes, that is basically what I wrote in UPGRADING, though if you feel I should explain it more clearly, I can try to do that.

The string you have shown (0xDE30D844) is invalid UTF-16, because the 2nd part of the surrogate pair is first and the first part is second. The previous implementation of `mb_strcut` runs all input strings through decoding/encoding filters, which convert such invalid byte sequences to error markers (`"?"`). The new implementation just examines a few bytes around each cut point to see if it looks like the cut point is falling in the middle of a character or not, and if so, adjusts it. The new implementation does not convert erroneous byte sequences to `?`.

That's why I said that the new implementation provides the same results as the old one for *valid* UTF-16 strings.

Please let me know what you think about this.

Reply:
@alexdowad Thank you for explain! 
I agree.
Looks good to me.